pkg/pushsync/pushsync.go

```diff
@@ -432,36 +383,42 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
 				}
 			}
 
-			ps.metrics.TotalSendAttempts.Inc()
+			// since we can reach into the neighborhood of the chunk
+			// act as the multiplexer and push the chunk in parallel to multiple peers
+			if ps.fullNode && swarm.Proximity(peer.Bytes(), ch.Address().Bytes()) >= ps.radiusChecker.StorageRadius() {
+				for ; parallelForwards > 0; parallelForwards-- {
+					retry()
+					sentErrorsLeft++
+				}
+			}
+
+			action, err := ps.prepareCredit(ctx, peer, ch, origin)
+			if err != nil {
+				ps.skipList.Add(ch.Address(), peer, overDraftRefresh)
+				retry()
+				continue
+			}
+			ps.skipList.Add(ch.Address(), peer, sanctionWait)
 
+			ps.metrics.TotalSendAttempts.Inc()
 			inflight++
 
-			go func() {
-				ctxd, cancel := context.WithTimeout(ctx, defaultTTL)
-				defer cancel()
-				ps.pushPeer(ctxd, skip, resultChan, done, peer, ch, origin)
-			}()
+			go ps.push(ctx, resultChan, done, peer, ch, action)
 
 		case result := <-resultChan:
 
 			inflight--
 
 			ps.measurePushPeer(result.pushTime, result.err, origin)
 
-			if ps.warmedUp() && !errors.Is(result.err, accounting.ErrOverdraft) {
-				ps.skipList.Add(ch.Address(), result.peer, sanctionWait)
-			}
-
 			if result.err == nil {
 				return result.receipt, nil
 			}
 
 			ps.metrics.TotalFailedSendAttempts.Inc()
 			logger.Debug("could not push to peer", "peer_address", result.peer, "error", result.err)
 
-			if result.sent {
-				sentErrorsLeft--
-			}
+			sentErrorsLeft--
 
 			retry()
```

Comment:
this case will process results of push operations and when first successful push has happens this function will return (line 415). when parallel forwards are also happening, it appears that there could be these issues:
1) logic from this case statement will not execute; which could skew metrics and telemetry a bit
2) race condition could prevent chunks to be forwarded to other peers if `pushToClosest` returns early after getting first successful push. this can happen because `context` and/or `doneChan` (whichever is relevant) would get closed therefore closing push goroutine.

Reply:
You have a valid point with the early context cancellation.

Reply:
I removed the done channel and the push result return now waits on the parent ctx's done channel instead.
The push func creates it's own context to allow each push attempt to run independently of the parent context

Reply:
for 1), wouldn't it be issue for `TotalSendAttempts` counter is incremented 5 times (1 default forward  + 4 multiplexing) and it's counterpart counter measuring errors `TotalFailedSendAttempts` never even when last 4 push operations have failed? It looks like there will be no logging and no telemetry for multiplexed chunks?